docs/devel/kubectl-conventions.md

```diff
@@ -0,0 +1,115 @@
+* kubectl VERB NOUNs for commands that apply to multiple resource types
+* NOUNs may be specified as TYPE name1 name2 ... or TYPE/name1 TYPE/name2; TYPE is omitted when only a single type is expected
+* Resource types are all lowercase, with no hyphens; both singular and plural forms are accepted
+* NOUNs may also be specified by one or more file arguments: -f file1 -f file2 ...
+* Resource types may have 2- or 3-letter aliases.
+* Business logic should be decoupled from the command framework, so that it can be reused independently of kubectl, cobra, etc.
+  * Ideally, commonly needed functionality would be implemented server-side in order to avoid problems typical of "fat" clients and to make it readily available to non-Go clients
+* Commands that generate resources, such as `run` or `expose`, should obey the following conventions:
+  * Flags should be converted to a parameter Go map or json map prior to invoking the generator
+  * The generator must be versioned so that users depending on a specific behavior may pin to that version, via `--generator=`
+  * Generation should be decoupled from creation
+  * `--dry-run` should output the resource that would be created, without creating it
+* A command group (e.g., `kubectl config`) may be used to group related non-standard commands, such as custom generators, mutations, and computations
+
+## Flag conventions
+
+* Flags are all lowercase, with words separated by hyphens
+* Flag names and single-character aliases should have the same meaning across all commands
+* Command-line flags corresponding to API fields should accept API enums exactly (e.g., --restart=Always)
+
+## Output conventions
+
+* By default, output is intended for humans rather than programs
+  * However, affordances are made for simple parsing of `get` output
+* Only errors should be directed to stderr
+* `get` commands should output one row per resource, and one resource per row
+  * Column titles and values should not contain spaces in order to facilitate commands that break lines into fields: cut, awk, etc.
+  * By default, `get` output should fit within about 80 columns
+    * Eventually we could perhaps auto-detect width
+    * `-o wide` may be used to display additional columns
+  * The first column should be the resource name, titled `NAME` (may change this to an abbreviation of resource type)
+  * NAMESPACE should be displayed as the first column when --all-namespaces is specified
+  * The last default column should be time since creation, titled `AGE`
+  * `-Lkey` should append a column containing the value of label with key `key`, with `<none>` if not present
+  * json, yaml, Go template, and jsonpath template formats should be supported and encouraged for subsequent processing
+    * Users should use --api-version or --output-version to ensure the output uses the version they expect
+* `describe` commands may output on multiple lines and may include information from related resources, such as events
```

Comment:
When fetching related resources, a targeted field selector should be used in favor of client side filtering of related resources.  

Have been fixing a few of our describe calls that fetch everything instead of a more targeted call.